### ISSUE TITLE
100% CPU usage after running "sonobuoy logs -f"

### ISSUE THREAD

**Issue by invidian** [vmware-tanzu/sonobuoy#1064]
**What steps did you take and what happened:**
After running even single test, e.g.: 
```
sonobuoy run --e2e-focus 'Aggregator Should be able to support the 1.10 Sample API Server using the current Aggregator'
```

And then leaving `sonobuoy logs -f` running for (in my case 2h), the process starts consuming 100% CPU of a single core.

**What did you expect to happen:**
`sonobuoy` to not trigger fans in my laptop :)

**Anything else you would like to add:**
`strace` doesn't say much, didn't try `gdb` or anything:
```
$ sudo strace -p 2554549
strace: Process 2554549 attached
futex(0x555c4acd9468, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x555c4acd9468, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x555c4acd9468, FUTEX_WAIT_PRIVATE, 0, NULL
```

If you have some suggestions for debugging, please let me know.

**Environment:**

- Sonobuoy version: 
```
$ sonobuoy version --kubeconfig local-testing/kubeconfig 
Sonobuoy Version: v0.17.0
MinimumKubeVersion: 1.15.0
MaximumKubeVersion: 1.17.99
GitSHA: 
API Version:  v1.17.0
```
- Kubernetes version: (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:20:10Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:12:17Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version: https://github.com/flexkube/libflexkube
- Cloud provider or hardware configuration: VirtualBox with Vagrant
- OS (e.g. from `/etc/os-release`): 
Kubernetes nodes:
```
core@controller01 ~ $ cat /etc/os-release 
NAME="Flatcar Container Linux by Kinvolk"
ID=flatcar
ID_LIKE=coreos
VERSION=2345.99.1
VERSION_ID=2345.99.1
BUILD_ID=2019-12-19-0008
PRETTY_NAME="Flatcar Container Linux by Kinvolk 2345.99.1 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://flatcar-linux.org/"
BUG_REPORT_URL="https://issues.flatcar-linux.org"
FLATCAR_BOARD="amd64-usr"
```
Host (where `sonobuoy` runs):
```
$ cat /etc/os-release 
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="0;36"
HOME_URL="https://www.archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux
```
- Sonobuoy tarball (which contains * below)

**Comment by zubron** [comment]
I can reproduce this issue locally though I haven't had a chance to look into it yet. Thanks for the report!

**Comment by johnSchnake** [comment]
@zubron How long did it take you to repro? I've had a test run going for a few min and `sonobuoy logs -f` doesnt seem to be consuming a lot of cpu at all (`top` is showing 0-0.5%).

**Comment by zubron** [comment]
@johnSchnake It was probably an hour or so. I left it running and forgot about it. Unfortunately didn't capture enough data yesterday but I was going to run it again today to gather more info. It was very clearly consuming one core at 100% after that time.

**Comment by zubron** [comment]
I spent some time time profiling and debugging this and have found the source of the issue:
**tl;dr**: The `ResultChan` in [`watchPodsToStreamLogs`](https://github.com/vmware-tanzu/sonobuoy/blob/master/pkg/client/logs.go#L170-L202) gets closed and since we don't handle the closed channel case, the loop is no longer blocked and we start an infinite loop.

More details below on how I found it. This was a fun one because I don't have a lot of experience in profiling so this was a great opportunity to learn :) I decided to share what I did here in case anyone else is curious!

I also tried using `strace` to see if I could collect any information there and realised that I also needed to use the `-f` flag to follow child processes/threads as I could see in `htop` that it was one of the threads that was maxing out. It had gone from a sleep state to running. The only thing of interest that I saw in the `strace` output was that the increase in CPU usage seemed to correspond to lots of continued calls to `nanosleep`:
```
[pid 28324] 22:36:47.155387 nanosleep({tv_sec=0, tv_nsec=10000000}, NULL) = 0
[pid 28324] 22:36:47.165525 nanosleep({tv_sec=0, tv_nsec=10000000}, NULL) = 0
```

I'm not an expert in using tools like `strace` so I didn't really know how to interpret what I was seeing. I then decided to modify the Sonobuoy client so that I could use [pprof](https://golang.org/pkg/net/http/pprof/). I profiled the program once the CPU usage hit 100% with the following command: `go tool pprof http://localhost:6060/debug/pprof/profile`. That collected a 30 second profile which I could then inspect with `go tool pprof -web ~/pprof/pprof.sonobuoy.samples.cpu.004.pb.gz` (where the filename is the one returned from the previous pprof command). That allowed me to look at the profile graph in my browser:

![profile001](https://user-images.githubusercontent.com/5786101/72073467-f5119780-32bd-11ea-83a7-2a9092091e58.png)

A profile from before the increase in CPU showed no nodes. All of the CPU usage was now being spent in the function [`watchPodsToStreamLogs`](https://github.com/vmware-tanzu/sonobuoy/blob/master/pkg/client/logs.go#L170-L202). Specifically, we are using the watching for events to see if any new pods are created, and using the [`ResultChan()`](https://github.com/kubernetes/apimachinery/blob/master/pkg/watch/watch.go#L34-L37) to watch for these. As we can see in the comment for that function, if an error occurs or `Stop()` is called, the channel will be closed. I'm not sure, but my guess is that the connection or request is closed after some time and that channel that we are using is closed. Given that closed channels don't block, and we are not specifically checking for a closed channel in that function, we start continuously looping reading from that closed channel which causes the increase in CPU usage.

You can trigger and inspect this locally by causing the connection to be dropped. For example, I started a quick run with Sonobuoy then used `sonobuoy logs -f` on a Kind cluster. Once the test was finished and Sonobuoy was idling, I deleted the cluster and noticed the quick increase in CPU usage since the channel was closed. I also came across this comment on an issue: https://github.com/kubernetes/client-go/issues/547#issuecomment-456442740 so it's up to to Sonobuoy to re-establish the watch for new events.

I hope someone finds this useful! It was a fun learning experience for me :)

**Comment by invidian** [comment]
Great findings @zubron! It can be reproduced using https://github.com/vzaliva/simpleproxy. I think fixing that should be simple, so I'll work on the patch.

**Comment by zubron** [comment]
That would be great, thanks @invidian!

**Comment by invidian** [comment]
Created draft PR addressing this issue: #1068. Please let me know what you think.

**Comment by Lirt** [comment]
Hi guys, this bug is pretty anoying, how are you doing with PR?

**Comment by zubron** [comment]
I'm sorry - I let this issue and the PR review fall off my radar. I'll try to get it addressed ASAP!

**Comment by invidian** [comment]
This issue is still relevant.

**Comment by vladimirvivien** [comment]
@invidian thanks for bringing this to our attention.
cc @wilsonehusin

**Comment by invidian** [comment]
Not stale

**Comment by vladimirvivien** [comment]
Thanks for keeping this open.
